### PR TITLE
Support BytesIO in read_image()

### PR DESCRIPTION
**Pull request by justinxzhao**
On the twitterbots dataset, training is failing when image features are enabled with this error AttributeError: '_io.BytesIO' object has no attribute 'shape'`.

This is caused by an issue with how `read_image()` doesn’t handle `BytesIO` objects returned by `get_image_from_http_bytes` when image features are URLs.

This PR fixes this, with minor cleanup to `read_image()`.

### CODE REVIEW

File: ludwig/utils/image_utils.py [hunk 1]
```diff
@@ -107,14 +108,35 @@ def is_image_score(src_path, img_entry):
...
+            image_tensor = decode_image(
+                torch.frombuffer(buffer_view, dtype=torch.uint8))
+            del(buffer_view)
+            return image_tensor
+    if isinstance(img, BytesIO):
+        buffer_view = img.getbuffer()
+        try:
+            image_tensor = decode_image(
+                torch.frombuffer(buffer_view, dtype=torch.uint8))
+            del(buffer_view)
```
Comment (skanjila) [inline ludwig/utils/image_utils.py:133]: is this basically memory cleanup?

File: ludwig/utils/image_utils.py [hunk 2]
```diff
@@ -107,14 +108,35 @@ def is_image_score(src_path, img_entry):
...
+    Returns None if the image could not be read.
     """
+    if isinstance(img, torch.Tensor):
+        return img
     if isinstance(img, str):
         return read_image_from_str(img, num_channels)
-    return img
+    if isinstance(img, bytes):
+        with BytesIO(img) as buffer:
+            buffer_view = buffer.getbuffer()
```
Comment (skanjila) [inline ludwig/utils/image_utils.py:123]: what happens if the img is too big to fit in the buffer?

File: ludwig/utils/image_utils.py [hunk 3]
```diff
@@ -107,14 +108,35 @@ def is_image_score(src_path, img_entry):
...
-    If num_channels is not provided, he image is read in unchanged format.
+    If num_channels is not provided, the image is read in unchanged format.
+    Returns None if the image could not be read.
     """
+    if isinstance(img, torch.Tensor):
+        return img
     if isinstance(img, str):
         return read_image_from_str(img, num_channels)
-    return img
+    if isinstance(img, bytes):
```
Comment (ShreyaR) [inline ludwig/utils/image_utils.py:121]: I believe @jimthompson5802's PR also adds some of these changes

**Review by ShreyaR — APPROVED** [review]
lgtm @justinxzhao! Jim's PR also adds a few of these changes in addition to tests for some image_utils functions, and it hasn't been merged yet. Feel free to merge if you can confirm that the tests pass!

File: ludwig/utils/image_utils.py (hunk 3 above)
Comment (justinxzhao) [reply ludwig/utils/image_utils.py:121]: Great observation!

I actually added my changes on top of Jim's MNIST fix. I'll wait for his PR to be merged first, and then rebase (and re-test) before merging.

File: ludwig/utils/image_utils.py (hunk 1 above)
Comment (justinxzhao) [reply ludwig/utils/image_utils.py:133]: I think so. I'm not sure if it's completely necessary to manually delete the buffer here. @jimthompson5802 what was your reasoning for calling `del` manually?

File: ludwig/utils/image_utils.py (hunk 2 above)
Comment (justinxzhao) [reply ludwig/utils/image_utils.py:123]: I imagine that the entire program would terminate with OOM. Hopefully no one's reading in images that are THAT big.

File: ludwig/utils/image_utils.py (hunk 1 above)
Comment (skanjila) [reply ludwig/utils/image_utils.py:133]: yes I was under the impression that garbage collector would kick in and do this for us, is that not the case?

**Pull request merged**